openmdao/utils/coloring.py

```diff
@@ -876,6 +965,232 @@ def on_resize(event):
+                  "object or by using 'openmdao view_coloring --textview <your_coloring_file>' "
+                  "from the command line.")
+            return
+
+        if isinstance(source, str):
+            coloring = Coloring.load(source)
+            source_name = pathlib.Path(source).absolute()
+        elif isinstance(source, Coloring):
+            coloring = source
+            source_name = ''
+        elif hasattr(source, '_coloring_info'):
+            coloring = source._coloring_info['coloring']
+            source_name = source._problem()._name
+        else:
+            raise ValueError(f'display_bokeh was expecting the source to be a valid coloring file '
+                             f'or an instance of driver but instead got f{type(source)}')
+
+        if coloring is None:
+            # Save and show
+            summary_div = PreText(text='No total derivative coloring found.',
+                                  styles={'font-size': '12pt'})
+
+            fig = PreText(text='')
+        else:
+            nrows, ncols = coloring._shape
+            aspect_ratio = ncols / nrows
+
+            tot_size, tot_colors, fwd_solves, rev_solves, pct = coloring._solves_info()
+
+            data = {}
+
+            # The row and column indices of the individual jacobian elements
+            data['col_idx'] = np.tile(np.arange(ncols, dtype=int), nrows)
+            data['row_idx'] = np.repeat(np.arange(nrows, dtype=int), ncols)
+
+            have_vars = None not in (coloring._col_vars, coloring._row_vars,
+                                     coloring._col_var_sizes, coloring._row_var_sizes)
+
+            # The indices of the responses and desvars obtained by binning the row/col indices
+            if have_vars:
+                desvar_idx_bins = [] if coloring._col_var_sizes is None else \
+                    np.cumsum(coloring._col_var_sizes)
+                response_idx_bins = [] if coloring._row_var_sizes is None else \
+                    np.cumsum(coloring._row_var_sizes)
+
+                response_idx = np.digitize(data['row_idx'], response_idx_bins)
+                desvar_idx = np.digitize(data['col_idx'], desvar_idx_bins)
+
+                data['pattern'] = np.full(nrows * ncols, '', dtype=str)
+                data['pattern'][...] = np.asarray(desvar_idx % 2 + response_idx % 2, dtype=str)
+            data['fwd_color_idx'] = np.full(nrows * ncols, '', dtype=object)
+            data['rev_color_idx'] = np.full(nrows * ncols, '', dtype=object)
+
+            # Add the color group information to the data source
+            fwd_map = {}
+            if coloring._fwd is not None:
+                for idx_fwd, (_cols, _nz_rows) in enumerate(coloring.color_nonzero_iter('fwd')):
+                    for _row_idx, _col_idx in zip(_nz_rows, _cols):
+                        fwd_grp_idxs = itertools.product(_row_idx, [_col_idx])
+                        fwd_map.update({(i, j): idx_fwd for i, j in fwd_grp_idxs})
```

Comment:
Since _col_idx is a single value, couldn't you just get rid of the product call and do `fwd_map.update({(i, _col_idx): idx_fwd for i in _row_idx})`?